Fix attach volume to instance repeatedly

FengyunPan opened the pull request:
1.When volume's status is 'attaching', controllermanager will attach
    it again and return err. So it is necessary to check volume's
    status before attach/detach volume.

   2. When volume's status is 'attaching', its attachments will be None,
    controllermanager can't get device path and make some failed event.
    But it is normal, so don't return err when attachments is None

Fix bug: #44536

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44537)
<!-- Reviewable:end -->

FengyunPan commented:
/release-note-none

derekwaynecarr commented:
is there a test case that can demonstrate the problem that was fixed?

FengyunPan commented:
@derekwaynecarr Hi, I have added a test case for it.

gnufied commented on pkg/volume/cinder/attacher.go:
```diff
@@ -72,6 +72,14 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
 	if err != nil {
 		return "", err
 	}
+	HasDone, VolumeStatus, err := attacher.cinderProvider.OperateHasDone(volumeID)
```
Fix the naming of these variables please. Make sure they start with lower case letters.

FengyunPan replied:
Done

gnufied commented on pkg/volume/cinder/attacher.go:
```diff
@@ -89,18 +97,27 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
+	// When volume's status is 'attaching', DiskIsAttached() return <false, nil>, and can't get device path.
+	// We should get device path next time and do not return err.
+	devicePath := ""
+	HasDone, _, err = attacher.cinderProvider.OperateHasDone(volumeID)
```
same as before about naming.

FengyunPan replied:
Done

FengyunPan replied:
Have updated it.

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -201,12 +206,34 @@ func (volumes *VolumesV2) deleteVolume(volumeName string) error {
 
+// return true when volume has been attached or detached.
+// or return false.
+func (os *OpenStack) OperateHasDone(diskName string) (bool, string, error) {
```
Can you rename this function to something like - `OperationPending` ?

FengyunPan replied:
Done.

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -74,8 +74,14 @@ type VolumeCreateOpts struct {
+	volumeInUseStatus = "in-use"
+	VolumeDeletedStatus = "deleted"
+	VolumeErrorStatus = "error"
+)
```
Lets be consistent about naming. Constants generally start with upper case letter.

FengyunPan replied:
Done.

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -201,12 +206,34 @@ func (volumes *VolumesV2) deleteVolume(volumeName string) error {
+	}
+	VolumeStatus := volume.Status
+	if (VolumeStatus == VolumeErrorStatus) {
+		glog.Errorf("volume's status is %s", VolumeStatus)
```
Can you log the disk name when we are logging the error? Without disk name, this error log will be probably useless

FengyunPan replied:
Done.

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -201,12 +206,34 @@ func (volumes *VolumesV2) deleteVolume(volumeName string) error {
 		return "", err
 	}
+	if volume.Status != volumeAvailableStatus {
+		errmsg := fmt.Sprintf("volume %s status is %s, not %s, can not be attached to instance %s.", volume.Name, volume.Status, volumeAvailableStatus, instanceID)
```
Please use ` fmt.Errorf`

FengyunPan replied:
the errmsg of errors.New(errmsg) should be a string not a error

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -378,7 +413,7 @@ func (os *OpenStack) GetAttachmentDiskPath(instanceID string, diskName string) (
 	}
-	return "", fmt.Errorf("volume %s is not attached to %s", diskName, instanceID)
+	return "", fmt.Errorf("volume %s has losed ServerId.", diskName)
 }
```
typo?

FengyunPan replied:
Done

idvoretskyi commented:
cc @kubernetes/sig-openstack-pr-reviews

dims commented:
@k8s-bot ok to test

dims commented:
@FengyunPan Can  you please check on the failed jobs?

Commit:
Don't return err when volume's status is 'attaching'

When volume's status is 'attaching', its attachments will be None,
controllermanager can't get device path and make some failed event.
But it is normal, let's fix it.

FengyunPan commented:
@k8s-bot bazel test this
@k8s-bot unit test this

Commit:
Fix reconciler test of attaching volume

The Attach() of FakeVolume should return device path

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

FengyunPan commented:
@k8s-bot kops aws e2e test this

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

FengyunPan commented:
@dims I have checked the failed jobs, we can ignore "pull-kubernetes-federation-e2e-gce" jobs because it is federation's jobs.

dims commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

dims commented:
/assign @dims

dims commented:
/approve
/lgtm

FengyunPan commented:
@jsafrane  @mikedanese Hi,  I'd like you to do a code review, thank you.

FengyunPan commented:
/assign @jsafrane @mikedanese

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jsafrane commented:
/approve

dims commented:
/assign @mikedanese

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

mikedanese commented:
/approve

saad-ali commented:
This PR broke Flex volumes. See https://github.com/kubernetes/kubernetes/issues/46882#issuecomment-306331570. Proposing reverting (at least partially this PR). Please discuss on that issue.
